exonum-java-binding-core/rust/src/utils/jni_cache.rs

```diff
@@ -0,0 +1,158 @@
+            "getStateHashes",
+            "(J)[[B",
+        );
+        USA_CACHED_CONVERT_TRANSACTION = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserServiceAdapter",
+            "convertTransaction",
+            "([B)Lcom/exonum/binding/service/adapters/UserTransactionAdapter;",
+        );
+        CLASS_CACHED_ERROR = env
+            .new_global_ref(env.find_class("java/lang/Error").unwrap().into())
+            .ok();
+        CLASS_CACHED_TEE = env
+            .new_global_ref(
+                env.find_class("com/exonum/binding/transaction/TransactionExecutionException")
+                    .unwrap()
+                    .into(),
+            ).ok();
+
+        assert!(
+            OBJECT_CACHED_GET_CLASS.is_some()
+                && CLASS_CACHED_ERROR.is_some()
+                && THROWABLE_CACHED_GET_MESSAGE.is_some()
+                && UTA_CACHED_EXECUTE.is_some()
+                && UTA_CACHED_INFO.is_some()
+                && UTA_CACHED_VERIFY.is_some()
+                && USA_CACHED_STATE_HASHES.is_some()
+                && USA_CACHED_CONVERT_TRANSACTION.is_some()
+                && CLASS_CACHED_ERROR.is_some()
+                && CLASS_CACHED_TEE.is_some(),
+            "Error caching Java entities"
+        );
+
+        info!("Done caching references to Java classes and methods.");
+    }
+}
+
+/// Helper method. Produces `JMethodID` for a particular class dealing with lifetime.
+fn get_method_id(env: &JNIEnv, class: &str, name: &str, sig: &str) -> Option<JMethodID<'static>> {
+    env.get_method_id(class, name, sig)
+        // we need this line to erase lifetime in order to save underlying raw pointer in static
+        .map(|mid| mid.into_inner().into())
+        .ok()
+}
+
+/// Returns cached `JClass` for `java/lang/Error` as a `GlobalRef`
+pub fn get_error_class() -> GlobalRef {
+    unsafe { CLASS_CACHED_ERROR.clone().unwrap() }
+}
+
+/// Returns cached `JClass` for `TransactionExecutionException` as a `GlobalRef`
+pub fn get_tee_class() -> GlobalRef {
+    unsafe { CLASS_CACHED_TEE.clone().unwrap() }
+}
+
+/// Returns cached `JMethodID` for `java.lang.Object.getClass()`
+pub fn get_object_get_class() -> JMethodID<'static> {
+    unsafe { OBJECT_CACHED_GET_CLASS.unwrap() }
+}
+
+/// Returns cached `JMethodID` for `java.lang.Class.getName()`
+pub fn get_class_get_name() -> JMethodID<'static> {
+    unsafe { CLASS_CACHED_GET_NAME.unwrap() }
+}
+
+/// Returns cached `JMethodID` for `java.lang.Throwable.getMessage()`
+pub fn get_throwable_get_message() -> JMethodID<'static> {
+    unsafe { THROWABLE_CACHED_GET_MESSAGE.unwrap() }
+}
+
+/// Returns cached `JMethodID` for `UserTransactionAdapter.execute()`
+pub fn get_uta_execute() -> JMethodID<'static> {
+    unsafe { UTA_CACHED_EXECUTE.unwrap() }
+}
+
+/// Returns cached `JMethodID` for `UserTransactionAdapter.info()`
+pub fn get_uta_info() -> JMethodID<'static> {
+    unsafe { UTA_CACHED_INFO.unwrap() }
```

Comment:
Is it possible to sprinkle some macros to remove the repeated code? Something like:

```rust
transaction_adapter {
  class_ref(UTA_CLASS_REF)
  method_id(verify, UTA_VERIFY_ID)
  method_id(execute, UTA_EXECUTE_ID)
  method_id(info, UTA_INFO_ID)
}
// becomes some four methods in a namespace?
```

@alekseysidorov, WDYT?

Reply:
Possible, but from my point of view macros bring a few major issues:
1. They are not supported by IDE well, and therefore it's impossible to use code navigation, refactoring tools and (in most cases) auto-completion
2. Macros are much harder to read and understand while briefly looking over code.
3. In case of any special needs macros will no more be an option.

Reply:
I also thought about this, but the initial requirement of this task was that change should not bring additional complexity to the existing code. I'd better implement such improvement as a separate task possibly with adding caching for some other entities. And yes, macros have pros and cons, to me, it looks like it worth trying if we need to cache more things.